sbx/recurrent_ppo/policies.py

```diff
@@ -0,0 +1,415 @@
+            cell_state
+        )
+
+        lstm_states = (hidden_state, cell_state)
+        hidden_size = lstm_states[0].shape[-1]
+        
+        new_lstm_states, output = nn.LSTMCell(features=hidden_size)(lstm_states, input)
+        return new_lstm_states, output
+    
+    @staticmethod
+    def initialize_carry(batch_size, hidden_size):
+        # Returns a tuple of lstm states (hidden and cell states)
+        return nn.LSTMCell(features=hidden_size).initialize_carry(
+            rng=jax.random.PRNGKey(0), input_shape=(batch_size, hidden_size)
+        )
+
+# Add ScanLSTM as first element of the Critic architecture
+class Critic(nn.Module):
+    n_units: int = 256
+    activation_fn: Callable[[jnp.ndarray], jnp.ndarray] = nn.tanh
+
+    # return hidden state + val
+    @nn.compact
+    def __call__(self, lstm_states, obs_dones) -> jnp.ndarray:
+        lstm_states, out = ScanLSTM()(lstm_states, obs_dones)
+        x = nn.Dense(self.n_units)(out)
+        x = self.activation_fn(x)
+        x = nn.Dense(self.n_units)(x)
+        x = self.activation_fn(x)
+        x = nn.Dense(1)(x)
+        return lstm_states, x
+
+# Add ScanLSTM as first element of the Actor architecture
+class Actor(nn.Module):
+    action_dim: int
+    n_units: int = 256
+    log_std_init: float = 0.0
+    activation_fn: Callable[[jnp.ndarray], jnp.ndarray] = nn.tanh
+    # For Discrete, MultiDiscrete and MultiBinary actions
+    num_discrete_choices: Optional[Union[int, Sequence[int]]] = None
+    # For MultiDiscrete
+    max_num_choices: int = 0
+    split_indices: np.ndarray = field(default_factory=lambda: np.array([]))
+
+    def get_std(self) -> jnp.ndarray:
+        # Make it work with gSDE
+        return jnp.array(0.0)
+
+    def __post_init__(self) -> None:
+        # For MultiDiscrete
+        if isinstance(self.num_discrete_choices, np.ndarray):
+            self.max_num_choices = max(self.num_discrete_choices)
+            # np.cumsum(...) gives the correct indices at which to split the flatten logits
+            self.split_indices = np.cumsum(self.num_discrete_choices[:-1])
+        super().__post_init__()
+
+    # return hidden state + action dist
+    @nn.compact
+    def __call__(self, hidden, obs_dones) -> tfd.Distribution:  # type: ignore[name-defined]
+        hidden, out = ScanLSTM()(hidden, obs_dones)
+        out = jnp.squeeze(out, axis=0)
+
+        x = nn.Dense(self.n_units)(out)
+        x = self.activation_fn(x)
+        x = nn.Dense(self.n_units)(x)
+        x = self.activation_fn(x)
+        action_logits = nn.Dense(self.action_dim)(x)
+        if self.num_discrete_choices is None:
+            # Continuous actions
+            log_std = self.param("log_std", constant(self.log_std_init), (self.action_dim,))
+            dist = tfd.MultivariateNormalDiag(loc=action_logits, scale_diag=jnp.exp(log_std))
+        elif isinstance(self.num_discrete_choices, int):
+            dist = tfd.Categorical(logits=action_logits)
+        else:
```

Comment:
to simplify, maybe consider only categorical and continuous action for now?

Reply:
Yes I checked and only those two action spaces work at the moment